parallaxinc/Parallax-IDE
# Automatic download

#### What's this PR do?

Adds automatic download functionality and logic to decide when to download. Address the issues linked below to deal with:
- No modules found
- 1 matching module found
- multiple matching modules found

Also adds a keypress for CTRL-R or F9 to automatically download code.
#### Where should the reviewer start?

Pull down this branch.
Get repo bs2-serial and pull down branch 'compiler-changes'
Get repo irken and pull down branch 'scan-options'
npm link to both repos
npm run build
#### How should this be manually tested?

Add some working code to your editor. With the correct directive and 1 matching STAMP plugged in, click the download button. The download window should pop up and automatically start downloading to the STAMP.

Try unplugging the STAMP. Download again, you should receive an error message as shown in the image below.

Try plugging in two STAMPS. You should get a message to pick one. Click a STAMP, it should automatically download without need for a Download button.

Try the CTRL-I and CTRL-R shortcut keys.
CTRL-I should Identify but not automatically download.
CTRL-R should bring up the download overlay and automatically download if a matching STAMP is found.
#### What are the relevant tickets?

Closes #131
Closes #163 
Closes #164 
Closes #165 
#### Screenshots

![image](https://cloud.githubusercontent.com/assets/2586086/8471202/be1f1944-204a-11e5-937e-fb3f6e7d5881.png)
